### DISCUSSION TITLE
Impact: Make PyKale more useful to public

### DISCUSSION THREAD

**Discussion by peizhenbai** [pykale/pykale#110] [General] [answered]
The discussion is for collecting ideas/resources about how we can efficiently develop PyKale and make it more useful to public, which includes but not limited to:
- Good resources for interdisciplinary/multimodal library development.
- Ideas about developing the following core API for graphs, texts, images and videos.
- Benchmark dataset collection and integration (e.g. [OGB](https://ogb.stanford.edu), [TDC](https://github.com/mims-harvard/TDC))
- Good workflow/pipeline for specific domains or popular research questions (e.g. [DeepPurpose](https://github.com/kexinhuang12345/DeepPurpose))

We can keep track of these good suggestions/issues and devote resources when feasible (e.g. BSc/MSc projects)

**Comment by bobturneruk** [comment]
Something of a "devil's advocate" / naïve question, but how doe PyKale compare to existing tools (e.g. scikit learn)? What niche(s) does it fill?

**Reply by haipinglu** [reply]
> Something of a "devil's advocate" / naïve question, but how doe PyKale

Welcome and great start! Compare our _neat_ [pipeline-based API](https://github.com/pykale/pykale/tree/master/kale) and the deep [scikit learn API](https://github.com/scikit-learn/scikit-learn/tree/main/sklearn). Which API is easier to read, use, and navigate? 

Our top-level API is pipeline-based, which is compact and will be rather stable because of the pipeline-based design. Scikit learn's top-level API is model-based so new module for each model, such as svm, nn. gp.

**Reply by haipinglu** [reply]
Actually, what @pz-white summarized in the bullets are mostly not offered elsewhere, at least in Scikit learn. How many checkboxes can Sckit learn check below?

- [ ] Good resources for interdisciplinary/multimodal library development.
- [ ] Ideas about developing the following core API for graphs, texts, images and videos.
- [ ] Benchmark dataset collection and integration (e.g. [OGB](https://ogb.stanford.edu), [TDC](https://github.com/mims-harvard/TDC))
- [ ] Good workflow/pipeline for specific domains or popular research questions (e.g. [DeepPurpose](https://github.com/kexinhuang12345/DeepPurpose))

**Reply by bobturneruk** [reply]
These look like substantial advantages and having them list will aid in advocacy for PyKale!

**Reply by peizhenbai** [reply]
Thanks for asking! Actually I'm considering how we can increase PyKale's competitiveness and let more people use it. Scikit learn is for the building of common machine learning models, and it is irrelevant to interdisciplinary/multimodal domain. Moreover, Scikit learn is highly integrated, which is convenient to apply common ML methods but hard to be used to develop researchers' own models.

I believe our Pykale is pipeline-based package and should be more flexible than Scikit. Now we have a good foundation of test framework and architecture, but it haven't reached the original aim (a interdisciplinary and multimodal ML-based package). We have different domains including graph, medical images and videos, but how we can build each domain well and let their components (e.g. model, dataset or evaluation) can be reproduced and inspire the other two.

Indeed it is a hard problem about combining different domains' knowledge. But from my experience, it is truly helpful. Like the attention mechanism, originally it is a techique applied in NLP but then rapidly extend to CV and graph to get a good performance. Sometimes the other domain's work is pretty meaningful.

Of course I think the current work focuses on how to build each domain well and then consider to combine together, so this discussion is open and welcome to share your ideas at any time. In furture, while acquiring resources, the ideas can give us many meaningful suggestions.

**Reply by haipinglu** [reply]
Well said. The above will be good for our future paper writing.

These ideas are good candidates of objectives for our [next project v0.2.0](https://github.com/pykale/pykale/projects/10). We need to break these plans into _actionable items_ to populate v0.2.0 objectives. This is extremely important, otherwise, even when we have the resources, we do not know what to do **specifically**.

@pz-white If you have some concrete, specific, and actionable steps already, even just small ones, you can start adding items to the list of [ v0.2.0](https://github.com/pykale/pykale/projects/10) or below, or we can start a new discussion about v0.2.0 objectives.

**Comment by bobturneruk** [comment]
It'd be interesting to identify groups of users (existing collaborators?) and maybe solicit feedback.

**Reply by haipinglu** [reply]
> It'd be interesting to identify groups of users (existing collaborators?) and maybe solicit feedback.

Yes, we are starting this process and I will promote more. But it will take time. We will grow from inside out.

**Comment by xianyuanliu** [comment, marked as answer]
We have secured a [publication](https://dl.acm.org/doi/10.1145/3511808.3557676) in CIKM and plan to integrate PyKale more extensively into [AIRE](https://shef-aire.github.io/project/) projects moving forward.
